File: src/memorygraph/utils/memory_parser.py
"""Shared memory parsing utilities."""

import json
import logging
from datetime import datetime
from typing import Any, Dict, Optional

from ..models import Memory, MemoryType, MemoryContext

logger = logging.getLogger(__name__)


def parse_memory_from_properties(
    node_data: Dict[str, Any],
    source: str = "unknown"
) -> Optional[Memory]:
    """
    Convert database node properties to Memory object.

    Works with Neo4j, FalkorDB, SQLite, and other backends.

    Args:
        node_data: Dictionary of node properties
        source: Source backend name for logging

    Returns:
        Memory object or None if parsing fails
    """
    try:
        # Extract basic memory fields
        memory_data = {
            "id": node_data.get("id"),
            "type": MemoryType(node_data.get("type")),
            "title": node_data.get("title"),
            "content": node_data.get("content"),
            "summary": node_data.get("summary"),
            "tags": node_data.get("tags", []),
            "importance": node_data.get("importance", 0.5),
            "confidence": node_data.get("confidence", 0.8),
            "effectiveness": node_data.get("effectiveness"),
            "usage_count": node_data.get("usage_count", 0),
            "created_at": _parse_datetime(node_data.get("created_at")),
            "updated_at": _parse_datetime(node_data.get("updated_at")),
        }

        # Handle optional last_accessed field
        if node_data.get("last_accessed"):
            memory_data["last_accessed"] = _parse_datetime(node_data["last_accessed"])

        # Extract context information
        context_data = _extract_context(node_data)
        if context_data:
            memory_data["context"] = MemoryContext(**context_data)

        return Memory(**memory_data)

    except Exception as e:
        logger.error(f"Failed to parse memory from {source}: {e}")
        return None


def _parse_datetime(value: Any) -> datetime:
    """Parse datetime from string or return as-is if already datetime."""
    if isinstance(value, str):
        return datetime.fromisoformat(value)
    return value


def _extract_context(node_data: Dict[str, Any]) -> Dict[str, Any]:
    """Extract context fields from node data."""
    context_data = {}
    for key, value in node_data.items():
        if key.startswith("context_") and value is not None:
            context_key = key[8:]  # Remove "context_" prefix

            # Deserialize JSON strings
            if isinstance(value, str) and context_key in ["additional_metadata"]:
                try:
                    context_data[context_key] = json.loads(value)
                except json.JSONDecodeError:
                    context_data[context_key] = value
            elif isinstance(value, str) and value.startswith(('[', '{')):
                try:
                    context_data[context_key] = json.loads(value)
                except json.JSONDecodeError:
                    context_data[context_key] = value
            else:
                context_data[context_key] = value

    # Handle timestamp fields in context
    if "timestamp" in context_data and isinstance(context_data["timestamp"], str):
        context_data["timestamp"] = datetime.fromisoformat(context_data["timestamp"])

    return context_data

File: ts/src/utils/memory-parser.ts
/**
 * Shared memory parsing utilities.
 *
 * Converts flat property dicts from graph DB backends into typed Memory objects.
 */

import {
  Memory,
  MemoryContext,
  MemoryType,
  isMemoryType,
  createMemory,
} from "../models.js";

/**
 * Convert database node properties to a Memory object.
 * Works with FalkorDB, Neo4j, SQLite, and other backends.
 */
export function parseMemoryFromProperties(
  nodeData: Record<string, unknown>,
  source = "unknown"
): Memory | null {
  try {
    const typeRaw = nodeData["type"] as string;
    const type = isMemoryType(typeRaw) ? typeRaw : MemoryType.GENERAL;

    const context = extractContext(nodeData);

    return createMemory({
      id: (nodeData["id"] as string) ?? null,
      type,
      title: (nodeData["title"] as string) ?? "",
      content: (nodeData["content"] as string) ?? "",
      summary: (nodeData["summary"] as string) ?? undefined,
      tags: (nodeData["tags"] as string[]) ?? [],
      importance: (nodeData["importance"] as number) ?? 0.5,
      confidence: (nodeData["confidence"] as number) ?? 0.8,
      effectiveness: (nodeData["effectiveness"] as number) ?? null,
      usage_count: (nodeData["usage_count"] as number) ?? 0,
      created_at: parseDateString(nodeData["created_at"]),
      updated_at: parseDateString(nodeData["updated_at"]),
      last_accessed: nodeData["last_accessed"]
        ? parseDateString(nodeData["last_accessed"])
        : undefined,
      context: context ?? undefined,
    });
  } catch (err) {
    console.error(`Failed to parse memory from ${source}: ${err}`);
    return null;
  }
}

function parseDateString(value: unknown): string {
  if (!value) return new Date().toISOString();
  if (value instanceof Date) return value.toISOString();
  return String(value);
}

function extractContext(nodeData: Record<string, unknown>): Partial<MemoryContext> | null {
  const contextData: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(nodeData)) {
    if (!key.startsWith("context_") || value === null || value === undefined) continue;
    const contextKey = key.slice(8); // remove "context_"

    // Deserialize JSON strings for known complex fields
    if (typeof value === "string") {
      if (
        contextKey === "additional_metadata" ||
        value.startsWith("[") ||
        value.startsWith("{")
      ) {
        try {
          contextData[contextKey] = JSON.parse(value);
        } catch {
          contextData[contextKey] = value;
        }
      } else {
        contextData[contextKey] = value;
      }
    } else {
      contextData[contextKey] = value;
    }
  }

  if (Object.keys(contextData).length === 0) return null;

  // Parse timestamp if present
  if (typeof contextData["timestamp"] === "string") {
    contextData["timestamp"] = new Date(contextData["timestamp"]).toISOString();
  }

  return contextData as Partial<MemoryContext>;
}
